### ISSUE TITLE
Oscillating val loss/acc with Batch Normalization.

### ISSUE THREAD

**Issue by jnphilipp** [fchollet/keras#3366]
I recently added batch normalization to my CNN. During training the validation loss and accuracy oscillate wildly. Has anyone an idea why?

```
    inputs = Input(shape=input_shape, name='input')
    x = inputs 
    for i in range(3): 
        x = Convolution2D(filters[0], 3, 3, border_mode='same', name='convolution2d%s_0' % i)(x) 
        x = BatchNormalization(axis=1)(x) 
        x = SReLU(name='srelu%s_0' % i)(x) 
        x = Convolution2D(filters[i], 3, 3, border_mode='same', name='convolution2d%s_1' % i)(x) 
        x = BatchNormalization(axis=1)(x) 
        x = SReLU(name='srelu%s_1' % i)(x) 
        x = Convolution2D(filters[i], 3, 3, border_mode='same', name='convolution2d%s_2' % i)(x) 
        x = BatchNormalization(axis=1)(x) 
        x = SReLU(name='srelu%s_2' % i)(x) 
        x = Convolution2D(filters[i], 3, 3, border_mode='same', subsample=(2, 2), name='convolution2d%s_3' % i)(x) 
        x = BatchNormalization(axis=1)(x) 
        x = SReLU(name='srelu%s_3' % i)(x) 

    x = Convolution2D(nb_classes, 4, 4, name='convolution2d3_0')(x) 
    x = BatchNormalization(axis=1)(x) 
    x = Flatten()(x)
    predictions = Activation('softmax')(x)

    model = Model(input=inputs, output=predictions)
    model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])

    train_datagen = ImageDataGenerator(
        rotation_range=30,
        width_shift_range=0.2,
        height_shift_range=0.2,
        rescale=1./255,
        shear_range=0.3,
        fill_mode='nearest')
    train_generator = train_datagen.flow(X, y, batch_size=batch_size, shuffle=True)

    test_datagen = ImageDataGenerator(rescale=1./255)
    validation_generator = test_datagen.flow(X, y, batch_size=batch_size)

    history = cnn.fit_generator(train_generator, len(X), nb_epoch, validation_data=validation_generator, nb_val_samples=len(X))
```

![train_history](https://cloud.githubusercontent.com/assets/1509101/17292797/6bd3d9c4-57ec-11e6-8a13-b4c29effb636.png)

**Comment by coljac** [comment]
I've noticed a similar issue. Did you have any resolution with this? 

`loss: 0.5240 - acc: 0.7566 - val_loss: 0.4623 - val_acc: 0.8093
loss: 0.3337 - acc: 0.8502 - val_loss: 0.4760 - val_acc: 0.8273
loss: 0.2761 - acc: 0.8756 - val_loss: 0.9010 - val_acc: 0.4381
loss: 0.2401 - acc: 0.8915 - val_loss: 0.6172 - val_acc: 0.5813
loss: 0.2218 - acc: 0.9006 - val_loss: 0.9548 - val_acc: 0.4723`

**Comment by the-moliver** [comment]
Off the top of my head I'd guess you have the wrong axis specified for your backend of choice. Tensorflow and Theano put the feature maps on different axes.

**Comment by isaacgerg** [comment]
I wonder if you have the momentum for the batch normalization too high for dataset so essentially it is batch dependent.

**Comment by gabriellabncsk** [comment]
Did anyone resolved this issue? I have the same problem, I have sudden jumps in validation loss.

**Comment by isaacgerg** [comment]
@gabriellabncsk Try lowering the "momentum" of your EMA in batch normalization.  I have to set my to 0.6 to get useful results.  Otherwise, the learned average of the normalization parameters is highly minibatch dependent.

**Comment by blaind** [comment]
Expanding on the-moliver's answer, check out the axis. Here's code for proper axis usage:

```
channel_axis = 1 if K.image_dim_ordering() == "th" else -1
model.add(BatchNormalization(axis=channel_axis))
```

**Comment by AhmedImtiazPrio** [comment]
Did anyone get a solution to this? I made sure that my batch is being normalized on the correct axis. I am using 1DCNN with a tensorflow backend, I have my axis specified as -1. As stated above, the validation accuracy and loss are oscillating wildly after adding batch normalization layers.

**Comment by angeldroth** [comment]
Anyone still looking for a solution to this, I found that lowering the layer momentum down to very low levels (I had to go ~0.01) removes the oscillation and keeps the loss and val loss in line with eachother.

**Comment by darylchang123** [comment]
I also observed oscillations in validation loss when adding batch norm before ReLU. We found that moving the batch norm after the ReLU resolved the issue.
